### PR TITLE
adding Gleam

### PR DESCRIPTION
**Pull request by chrislusf**
first draft

### CODE REVIEW

File: content/advent-2016/gleam-distributed-map-reduce-for-golang.md [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+
+Then I found LuaJIT. It seems an ideal complement to Go. For many who do
+not know LuaJIT, its performance is on the same level as C, Java, Go. Being
+a scripting language, it perfectly resolves the two limitations above:
+generics and dynamic remote code execution. Lua was created as a small and
+embeddable language. Go should embrace it. (BTW, LuaJIT's FFI is really
+easy to call an external C library function, much simpler than CGO.)
+
+Additionally, Unix Pipe tools, e.g., "sort", "tr", "uniq", etc, 
+are also gems for data proccesing. But being single threaded, they lack
```
Comment (bradleyfalzon) [inline content/advent-2016/gleam-distributed-map-reduce-for-golang.md]: `proccesing` should be `processing`

File: content/advent-2016/gleam-distributed-map-reduce-for-golang.md [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+the power to tackle big data. Go should be able to help to scale them up,
+to distributedly run these tools.
+
+With [Gleam](https://github.com/chrislusf/gleam), we can combine pure Go, 
+LuaJIT, and Unix Pipes, all the 3 powerful weapons together.
+
+Let's understand Gleam architecutre of how it works. 
+Then I will cover 2 examples:
+
+1. Sort an 1GB file in both standalone and distributed modes, and distributed
```
Comment (bradleyfalzon) [inline content/advent-2016/gleam-distributed-map-reduce-for-golang.md]: `Sort an` should be `Sort a`

File: content/advent-2016/gleam-distributed-map-reduce-for-golang.md [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+
+Additionally, Unix Pipe tools, e.g., "sort", "tr", "uniq", etc, 
+are also gems for data proccesing. But being single threaded, they lack 
+the power to tackle big data. Go should be able to help to scale them up,
+to distributedly run these tools.
+
+With [Gleam](https://github.com/chrislusf/gleam), we can combine pure Go, 
+LuaJIT, and Unix Pipes, all the 3 powerful weapons together.
+
+Let's understand Gleam architecutre of how it works.
```
Comment (bradleyfalzon) [inline content/advent-2016/gleam-distributed-map-reduce-for-golang.md]: `architecutre` should be `architecture`

File: content/advent-2016/gleam-distributed-map-reduce-for-golang.md [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+
+With [Gleam](https://github.com/chrislusf/gleam), we can combine pure Go, 
+LuaJIT, and Unix Pipes, all the 3 powerful weapons together.
+
+Let's understand Gleam architecutre of how it works. 
+Then I will cover 2 examples:
+
+1. Sort an 1GB file in both standalone and distributed modes, and distributed
+Unix sort mode, with performance comparison.
+2. Implement Joining of CSV files.
```
Comment (bradleyfalzon) [inline content/advent-2016/gleam-distributed-map-reduce-for-golang.md]: `Joining` should be `joining`, or at least is there is a reason it's capitalised?

File: content/advent-2016/gleam-distributed-map-reduce-for-golang.md [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+
+The flow can also be executed in a distributed cluster. The Gleam cluster
+has (currently) one master and multiple agents. The master's only job is
+to collect resource statuses from agents. And agents needs to report their 
+statuses to the master.
+
+The agents also manage intermediate data. The data can be streamed via memory,
+or optionally go through disk for better scalability and robustness.
+
+When run in a distributed cluster, the Gleam code becomes a driver of
```
Comment (bradleyfalzon) [inline content/advent-2016/gleam-distributed-map-reduce-for-golang.md]: I think `When run` should be `When ran`

File: content/advent-2016/gleam-distributed-map-reduce-for-golang.md [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+has (currently) one master and multiple agents. The master's only job is
+to collect resource statuses from agents. And agents needs to report their 
+statuses to the master.
+
+The agents also manage intermediate data. The data can be streamed via memory,
+or optionally go through disk for better scalability and robustness.
+
+When run in a distributed cluster, the Gleam code becomes a driver of 
+the whole flow's execution. It requests resources from the master, and then
+contact the agents to schedule jobs on them.
```
Comment (bradleyfalzon) [inline content/advent-2016/gleam-distributed-map-reduce-for-golang.md]: I think `contact` should be `contacts`

File: content/advent-2016/gleam-distributed-map-reduce-for-golang.md [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+sys	0m1.309s
+```
+Peak Memory used: 1.46 GB, CPU is about 100%.
+
+## Gleam Standalone mode
+
+The complete source code is here. It has these steps:
+
+1. read the input txt file line by line
+- extract the interger part as a string from each line (LuaJIT)
```
Comment (bradleyfalzon) [inline content/advent-2016/gleam-distributed-map-reduce-for-golang.md]: `interger` should be `integer`

File: content/advent-2016/gleam-distributed-map-reduce-for-golang.md [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+- hash the key and distribute data into 4 partitions
+- sort the key, first locally, then merge then into one partition
+- print out data to stdout.
+
+There could be better algorithms. For example, instead of randomly hash the
```
Comment (bradleyfalzon) [inline content/advent-2016/gleam-distributed-map-reduce-for-golang.md]: I think `hash` should be `hashing`

File: content/advent-2016/gleam-distributed-map-reduce-for-golang.md [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+
+## Gleam Distributed mode with Unix Sort
+
+Next example needs a Gleam cluster. See 
+https://github.com/chrislusf/gleam/wiki/Gleam-Cluster-Setup for instructions.
+
+This piece of code does the following:
+
+1. read the input txt file line by line
+2. extract the interger part as a string from each line (LuaJIT)
```
Comment (bradleyfalzon) [inline content/advent-2016/gleam-distributed-map-reduce-for-golang.md]: `interger` should be `integer`

File: content/advent-2016/gleam-distributed-map-reduce-for-golang.md [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+The total time is about the same as the standalone Gleam sort code.
+
+The logs are just scheduling information, showing the tasks to do,
+the amount of memory needed, the resources allocated, etc. Next example will
+show a way to adjust the required memory.
+
+The data flowing between each dataset are usually in MessagePack format.
+But the data flowing to and from Pipe() are tab-separated lines, 
+when each line has multiple fields. So there is a small amount of time 
+for the extra convertion.
```
Comment (bradleyfalzon) [inline content/advent-2016/gleam-distributed-map-reduce-for-golang.md]: `convertion` should be `conversion`

File: content/advent-2016/gleam-distributed-map-reduce-for-golang.md [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+
+As you can see, as long as any program can work as Unix Pipes, Gleam
+can make them work distributedly.
+
+## Gleam Distributed mode with Gleam Sort
+
+This piece of code takes these steps:
+
+1. read the input txt file line by line
+2. extract the interger part as a string from each line (LuaJIT)
```
Comment (bradleyfalzon) [inline content/advent-2016/gleam-distributed-map-reduce-for-golang.md]: `interger` should be `integer`

File: content/advent-2016/gleam-distributed-map-reduce-for-golang.md [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+    a := f.ReadFile(csv.New("a.csv")).Select(Field(1,4)) // a1, a4
+    b := f.ReadFile(csv.New("b.csv")).Select(Field(2,3)) // b2, b3
+
+    a.Join(b).Fprintf(os.Stdout, "%s,%s,%s\n").Run()  // a1, a4, b3
+
+}
+
+```
+In the example, the csv files are on local disk. This is the simplest form. 
+The csv files can also be read from HDFS or S3.
```
Comment (bradleyfalzon) [inline content/advent-2016/gleam-distributed-map-reduce-for-golang.md]: nit: I think `csv` should be `CSV`, this keeps consistent with `HDFS` and the next paragraph's use of `CSV`

File: content/advent-2016/gleam-distributed-map-reduce-for-golang.md [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+
+CSV is one of the storage format supported. There are different adapters 
+for different data sources, e.g. Gleam can read from Cassandra in parallel.
+More adapters are planned.
+
+# Summary
+
+Gleam is a simple and powerful distributed map reduce system.
+By combining high performance LuaJIT and Go's powerful system programing,
```
Comment (bradleyfalzon) [inline content/advent-2016/gleam-distributed-map-reduce-for-golang.md]: `programing` should be `programming`

**Comment by chrislusf** [comment]
Thanks for your careful reading! Fixed all accordingly!

**Comment by chrislusf** [comment]
I am wondering why this is not published today. And do I need to change the date in the post accordingly?

**Comment by chrislusf** [comment]
@dgryski This is the article for "Go+LuaJIT for efficient distributed map reduce system" scheduled today. I wonder why it is not used? Thanks!

**Comment by chrislusf** [comment]
@dgryski I am following https://docs.google.com/spreadsheets/d/1shwS1o-D2fPihea1NjKDH9KMheteRaJEqBjgsxkGmZA/edit#gid=0

**Comment by dgryski** [comment]
Indeed. I appear to have gotten off by one day somewhere. Your article will be published tomorrow.

Sorry for the confusion.

**Comment by chrislusf** [comment]
Thanks for your fast reply!

Just let me know whether I need to adjust the "date" metadata in the post.

**Comment by dgryski** [comment]
I will fix it when I merge

**Comment by dgryski** [comment]
This is now live.  Thanks for your contribution!

**Pull request merged**